src/spurt/graph/_delaunay.py

```diff
@@ -61,8 +67,115 @@ def boundary(self) -> np.ndarray:
+        pts = np.zeros((self.npoints, 2), dtype=int)
+        for ii in range(self._shape[0]):
+            i0 = ii * self._shape[1]
+            i1 = i0 + self._shape[1]
+
+            pts[i0:i1, 0] = np.arange(self._shape[1])
+            pts[i0:i1, 1] = ii
+
+        return pts
+
+    @property
+    def cycles(self) -> np.ndarray:
+        """Return rectangular loops in 2D grid."""
+        ncyc = self.npoints - self._shape[0] - self._shape[1] + 1
+        cyc = np.zeros((ncyc, 4), dtype=int)
+
+        ind = 0
+        # Iterate over the loops for top-left corner of loop
+        for ii in range(self._shape[0] - 1):
+            for jj in range(self._shape[1] - 1):
+                # cycles will go counter-clockwise like Delaunay
+                # top-left -> bottom-left -> bottom-right -> top-right
+                cyc[ind, :] = np.ravel_multi_index(
+                    ((ii, ii + 1, ii + 1, ii), (jj, jj, jj + 1, jj + 1)), self._shape
+                )
+                ind += 1
+
+        return cyc
+
+    @property
+    def boundary(self) -> np.ndarray:
+        """Return boundary edges in 2D grid."""
+        narcs = 2 * (self._shape[0] + self._shape[1] - 2)
+        arcs = np.zeros((narcs, 2), dtype=int)
+
+        ind = 0
+        # Walk along the top edge
+        for ii in range(self._shape[1] - 1):
+            arcs[ind, :] = (ii, ii + 1)
+            ind += 1
+
+        # Walk down the right edge
+        for ii in range(self._shape[0] - 1):
+            off = ii * self._shape[1] + self._shape[1] - 1
+            arcs[ind, :] = (off, off + self._shape[1])
+            ind += 1
+
+        # Walk back along bottom edge
+        for ii in range(self._shape[1] - 1, 0, -1):
+            off = (self._shape[0] - 1) * self._shape[1]
+            arcs[ind, :] = (off + ii - 1, off + ii)
+            ind += 1
+
+        # Walk back along left edge
+        for ii in range(self._shape[0] - 1, 0, -1):
+            off = ii * self._shape[1]
+            arcs[ind, :] = (off - self._shape[1], off)
+
+        return arcs
+
+    @property
+    def links(self) -> np.ndarray:
+        """Horizontal links followed by vertical links."""
+        narcs = 2 * self.npoints - self._shape[0] - self._shape[1]
+        arcs = np.zeros((narcs, 2), dtype=int)
+
+        ind = 0
+        # Start with horizontal links, iterate over the rows
+        for ii in range(self._shape[0]):
+            start = ii * self._shape[1] + np.arange(self._shape[1] - 1)
+            arcs[ind : ind + start.size, 0] = start
+            arcs[ind : ind + start.size, 1] = start + 1
+            ind += start.size
+
+        # Now include the vertical links, but walk along rows
+        for ii in range(self._shape[0] - 1):
+            start = ii * self._shape[1] + np.arange(self._shape[1])
+            arcs[ind : ind + start.size, 0] = start
+            arcs[ind : ind + start.size, 1] = start + self._shape[1]
+            ind += start.size
+
+        return arcs
```

Comment:
The resulting array only includes arcs within the 2-D grid of nodes. It doesn't include any arcs between boundary nodes and the "ground" node.

Is that intentional or do we need to also include the off-grid arcs here as well?

Reply:
The grounding node is a node in the dual graph. This would corresponding to a grounding cycle in the 2-D grid. This can also be thought of just walking along the boundary and making one big loop outside of the 2-D graph - and hence does not need additional edges